File: 733-flood-fill/733-flood-fill.rb
# @param {Integer[][]} image
# @param {Integer} sr
# @param {Integer} sc
# @param {Integer} color
# @return {Integer[][]}

def in_bounds(r, c, image)
    return r >= 0 && r < image.length && c >= 0 && c < image[0].length
end

def valid(r, c, image, visited, scolor)
    return in_bounds(r, c, image) && !visited.include?([r, c]) && image[r][c] == scolor
end

def dfs(r, c, image, visited, color)
    dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]]
    scolor = image[r][c]
    image[r][c] = color
    visited.add([r, c])
    dirs.each do |dr, dc|
        new_row, new_col = r + dr, c + dc
        if valid(new_row, new_col, image, visited, scolor)
            dfs(new_row, new_col, image, visited, color)
        end
    end
end

def flood_fill(image, sr, sc, color)
    visited = Set.new
    dfs(sr, sc, image, visited, color)
    return image
end

File: 733-flood-fill/733-flood-fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} color
 * @return {number[][]}
 */
var floodFill = function(image, sr, sc, color) {
    const [nRows, nCols] = [image.length, image[0].length];
    const dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    const visited = new Set();
    
    const inBounds = (row, col) => {
        return row >= 0 && row < nRows && col >= 0 && col < nCols;
    };
    
    const dfs = (row, col) => {
        const scolor = image[row][col];
        
        const valid = (r, c) => {
            return inBounds(r, c) && !visited.has(`${r}-${c}`) && image[r][c] === scolor;
        };
        
        image[row][col] = color;
        visited.add(`${row}-${col}`);
        dirs.forEach(([rowDiff, colDiff]) => {
            const [newRow, newCol] = [row + rowDiff, col + colDiff];
            if (valid(newRow, newCol)) dfs(newRow, newCol);
        });
    };
    
    dfs(sr, sc);
    return image;
};